src/main.rs

```diff
@@ -0,0 +1,270 @@
+use fastly::{downstream_client_ip_addr, uap_parse, Body, Error, Request, Response, ResponseExt};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::io::Write;
+use std::net::IpAddr;
+use std::str::FromStr;
+
+/// ReportBody models the body of a Network Error Log report which details the
+/// network error that occurred.
+///
+/// Note: view the Network Error Logging [specification](https://www.w3.org/TR/network-error-logging]
+/// for detailed information on the report structure.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct ReportBody {
+    #[serde(rename = "type")]
+    pub error_type: String,
+    pub elapsed_time: i32,
+    pub method: String,
+    pub phase: String,
+    pub protocol: String,
+    pub referrer: String,
+    pub sampling_fraction: f32,
+    pub server_ip: String,
+    pub status_code: i32,
+}
+
+/// Report models a NEL report, a collection of arbitrary data which a user agent
+/// is expected to deliver to the report endpoint.
+///
+/// Each report has a `body`, which is either null or an object which can be
+/// serialized into a JSON text. The fields contained in a report’s body are
+/// determined by the report’s type.
+///
+/// Each report has a `url`, which is typically the address of the Document or
+/// Worker from which the report was generated.
+///
+/// Each report has a `user_agent`, which is the value of the User-Agent header
+/// of the request from which the report was generated.
+///
+/// Each report has a `type`, which is a `ReportType`.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Report {
+    pub user_agent: String,
+    pub url: String,
+    #[serde(rename = "type")]
+    pub report_type: String,
+    pub body: ReportBody,
+    pub age: i64,
+}
+
+/// Parses an IP string input and truncates it to a privacy safe prefix mask and
+/// returns the network as a CIDR string, such as `167.98.105.176/28`.
+///
+/// For IPv4 addresses we truncate to a /28 prefix and for IPv6 addresses we
+/// truncate to /56.
+pub fn truncate_ip_to_prefix(ip: IpAddr) -> Result<String, Error> {
+    match ip {
+        IpAddr::V4(ip) => ipnet::Ipv4Net::new(ip, 28)
+            .map(|i| i.trunc().to_string())
+            .map_err(Error::new),
+        IpAddr::V6(ip) => ipnet::Ipv6Net::new(ip, 56)
+            .map(|i| i.trunc().to_string())
+            .map_err(Error::new),
+    }
+}
+
+/// UserAgent is a structured representation of a User Agent string which lets
+/// servers and network peers identify the application, operating system,
+/// vendor, and/or version of the requesting user agent.
+///
+/// Implements the `FromStr` trait to facilitate parsing from a User-Agent header
+/// value.
+#[derive(Clone)]
+pub struct UserAgent {
+    family: String,
+    major: String,
+    minor: String,
+    patch: String,
+}
+
+impl FromStr for UserAgent {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<UserAgent, Error> {
+        let (family, major, minor, patch) = uap_parse(s)?;
```

Comment:
Do we want to bee showing casing the `uap_parse` method publicly, when it will soon be replaced by the device detection API?